### PR TITLE
Zernicke poly dde

### PR DESCRIPTION
**Pull request by JoshVStaden**
- [ ] Tests added / passed

  ```bash
  $ py.test -v -s africanus
  ```

  If the pycodestyle tests fail, the quickest way to correct
  this is to run `autopep8` and then `pycodestyle` to fix the
  remaining issues.

  ```
  $ pip install -U autopep8 pycodestyle
  $ autopep8 -r -i africanus
  $ pycodestyle africanus
  ```

- [ ] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by sjperkins** [comment]
Reproducing email comments by @landmanbester  

> From what I can gather you plan to evaluate the polynomials as a function of p and phi (the radial distance and azimuthal angle). However, the source coordinates that get passed in are in (l,m) so you will have to implement the coordinate transformation (no need to pass in both coords and p, phi). I think it's as simple as switching between Euclidean and polar coordinates. Maybe its time to have a look at how Khan's polynomials are implemented. Oleg do you have the link to the package or should we ask Khan?

**Comment by sjperkins** [comment]
OK so at this point we're lacking an implementation. @JoshVStaden should figure this out,  checked by @landmanbester. It would be great if @kmbasad could eyeball the final version.

**Comment by JoshVStaden** [comment]
@sjperkins Hi Simon, I have put the Zernicke polynomial code in. I'm not entirely sure if I handled the input coords correctly. Please advise.

**Comment by sjperkins** [comment]
Please add nopython and nogil to the jit decorator. Cut and pasting the code is likely not going to work out the box.

**Comment by sjperkins** [comment]
I would also try running a test case with actual inputs to see if the function works, which I don't think it does

**Comment by sjperkins** [comment]
> @sjperkins am I on the right track here?

@JoshVStaden I don't think so. It looks like you've copied the code for creating the Zernicke Polynomials, whereas `zernicke_dde` should be reconstructing an `image` (in this case a 4D `(nsrc,ntime,nchan,ncorr)` image). It looks like this happens [here](https://github.com/kmbasad/eidos/blob/master/eidos/zernike.py#L198) in the eidos code. You'll need to convert this to the [Functional Programming](https://en.wikipedia.org/wiki/Functional_programming) style used in africanus.

**Comment by sjperkins** [comment]
Having said that, it'd probably be useful to have the code for creating the Zernicke polynomials in Africanus, especially for creating a test case.

**Comment by sjperkins** [comment]
@JoshVStaden  I think the docstring and outline of the `zernicke_dde` function should looks as follows:

```python
@numba.jit(nopython=True, nogil=True, cache=True)
def zernicke_dde(coords, coeffs):
"""
Evaluate Zernicke Polynomials defined by coefficients ``coeffs`
at the specified coordinates ``coords``.

Parameters
---------------
coords : :class:`numpy.ndarray`
   Coordinates at which to evaluate the zernicke polynomials.
   Has shape :code:`(3, source, time, ant, chan)`. The three components in
   the first dimension represent l, m and frequency coordinates, respectively.
coeffs : :class:`numpy.ndarray`
  Zernicke polynomial coefficients.
  Has shape :code:`(ant, chan, poly)` where ``poly`` is the number of
  polynomical coefficients.

Returns
----------
:class:`numpy.ndarray`
   complex values with shape :code:`(source, time, ant, chan)`

"""

_, sources, times, ants, chans = coords.shape

result = np.empty((sources,time,ants,chans), np.complex128)

for s in range(sources):
    for t in range(times):
        for a in range(ants):
            for c in range(chans):
                l, m, freq = coords[:,s,t,a,c]
                zcoeff = coeffs[a,c]
                # Insert implementation here
                result[s,t,a,c] = ...

return result
```

**Comment by sjperkins** [comment]
@JoshVStaden  I would also suggest installing and playing with [eidos](https://github.com/kmbasad/eidos). @kmbasad added a 2.34MB [coefficients](https://github.com/kmbasad/eidos/blob/master/eidos/data/meerkat_coeff_dict.npy) file. From my reading of the code, this can be used to reconstruct the full beam cubes. I would suggest experimenting with eidos to get a sense for how decomposition into coefficients and reconstruction from coefficients works.

**Comment by JoshVStaden** [comment]
Thanks Simon

I'm having a few issues installing Eidos. It seems that it was  
programmed for python 2.7 (hence the format of print is print  
"statement" instead of print("statement")), however, when I run the  
install with python 2.7, it sends an error message stating that I need  
python 3 to install it. Not sure what needs to be done in this case.

Quoting Simon Perkins <notifications@github.com>:

**Comment by sjperkins** [comment]
@JoshVStaden The following works for me with python 2.7.12 on Ubuntu 16.04:

```
sperkins@simon-Y700:~/work/ska/code$ git clone https://github.com/kmbasad/eidos
Cloning into 'eidos'...
remote: Counting objects: 127, done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 127 (delta 7), reused 13 (delta 6), pack-reused 106
Receiving objects: 100% (127/127), 1.95 MiB | 1.07 MiB/s, done.
Resolving deltas: 100% (59/59), done.
Checking connectivity... done.
sperkins@simon-Y700:~/work/ska/code$ cd eidos/
sperkins@simon-Y700:~/work/ska/code/eidos$ ls
bin  Dockerfile  eidos  README.md  setup.py
sperkins@simon-Y700:~/work/ska/code/eidos$ virtualenv ~/venv/eidos
New python executable in /home/sperkins/venv/eidos/bin/python
Installing setuptools, pip, wheel...done.
sperkins@simon-Y700:~/work/ska/code/eidos$ source ~/venv/eidos/bin/activate
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ pip install -e .
Obtaining file:///home/sperkins/work/ska/code/eidos
Collecting lmfit>=0.1.0 (from eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/7e/ea/88eeff9026d5b6a4063c4ba75d917b7b231a977fe258a29758411ba6f661/lmfit-0.9.9.tar.gz (1.6MB)
    100% |████████████████████████████████| 1.6MB 4.4MB/s 
Collecting numpy>=1.13.1 (from eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/c0/e7/08f059a00367fd613e4f2875a16c70b6237268a1d6d166c6d36acada8301/numpy-1.14.3-cp27-cp27mu-manylinux1_x86_64.whl (12.1MB)
    100% |████████████████████████████████| 12.1MB 2.3MB/s 
Collecting scipy>=0.19.1 (from eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/2a/f3/de9c1bd16311982711209edaa8c6caa962db30ebb6a8cc6f1dcd2d3ef616/scipy-1.1.0-cp27-cp27mu-manylinux1_x86_64.whl (30.8MB)
    100% |████████████████████████████████| 30.8MB 1.3MB/s 
Collecting astropy>=0.19.1 (from eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/32/81/ddad68500257b16f5e309c9bb81ddabdf67d4dd72a86c3f4b8ee728d2468/astropy-2.0.6-cp27-cp27mu-manylinux1_x86_64.whl (6.1MB)
    100% |████████████████████████████████| 6.1MB 6.8MB/s 
Collecting future (from eidos==0.1.0)
Collecting six (from lmfit>=0.1.0->eidos==0.1.0)
  Using cached https://files.pythonhosted.org/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Collecting asteval (from lmfit>=0.1.0->eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/6f/6b/63b958aef9e2b155450d3059756e68d54ead0898b67a793a7ab79ccbbabd/asteval-0.9.12.tar.gz (50kB)
    100% |████████████████████████████████| 51kB 11.2MB/s 
Collecting pytest>=2.8 (from astropy>=0.19.1->eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/76/52/fc48d02492d9e6070cb672d9133382e83084f567f88eff1c27bd2c6c27a8/pytest-3.5.1-py2.py3-none-any.whl (192kB)
    100% |████████████████████████████████| 194kB 15.0MB/s 
Collecting py>=1.5.0 (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0)
  Using cached https://files.pythonhosted.org/packages/67/a5/f77982214dd4c8fd104b066f249adea2c49e25e8703d284382eb5e9ab35a/py-1.5.3-py2.py3-none-any.whl
Requirement already satisfied: setuptools in /home/sperkins/venv/eidos/lib/python2.7/site-packages (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0) (39.1.0)
Collecting more-itertools>=4.0.0 (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0)
  Using cached https://files.pythonhosted.org/packages/4a/88/c28e2a2da8f3dc3a391d9c97ad949f2ea0c05198222e7e6af176e5bf9b26/more_itertools-4.1.0-py2-none-any.whl
Collecting funcsigs; python_version < "3.0" (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0)
  Using cached https://files.pythonhosted.org/packages/69/cb/f5be453359271714c01b9bd06126eaf2e368f1fddfff30818754b5ac2328/funcsigs-1.0.2-py2.py3-none-any.whl
Collecting attrs>=17.4.0 (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0)
  Downloading https://files.pythonhosted.org/packages/41/59/cedf87e91ed541be7957c501a92102f9cc6363c623a7666d69d51c78ac5b/attrs-18.1.0-py2.py3-none-any.whl
Collecting pluggy<0.7,>=0.5 (from pytest>=2.8->astropy>=0.19.1->eidos==0.1.0)
  Using cached https://files.pythonhosted.org/packages/82/05/43e3947125a2137cba4746135c75934ceed1863f27e050fc560052104a71/pluggy-0.6.0-py2-none-any.whl
Building wheels for collected packages: lmfit, asteval
  Running setup.py bdist_wheel for lmfit ... done
  Stored in directory: /home/sperkins/.cache/pip/wheels/da/b5/08/57c7b67e0198b18d53c51a200943d545339d8aa5faf91f2d6e
  Running setup.py bdist_wheel for asteval ... done
  Stored in directory: /home/sperkins/.cache/pip/wheels/8c/33/aa/c68610f385996f1eed253442e7ab0dbe4046c51f560ca7abd9
Successfully built lmfit asteval
Installing collected packages: numpy, scipy, six, asteval, lmfit, py, more-itertools, funcsigs, attrs, pluggy, pytest, astropy, future, eidos
  Running setup.py develop for eidos
Successfully installed asteval-0.9.12 astropy-2.0.6 attrs-18.1.0 eidos funcsigs-1.0.2 future-0.16.0 lmfit-0.9.9 more-itertools-4.1.0 numpy-1.14.3 pluggy-0.6.0 py-1.5.3 pytest-3.5.1 scipy-1.1.0 six-1.11.0
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ eidos
usage: eidos [-h] -p PIXELS [-d DIAMETER] -f FREQ [FREQ ...]
             [-c COEFFICIENTS_FILE] [-P PREFIX] [-o8] [-N]
eidos: error: argument -p/--pixels is required
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ eidos -c 
bin/            eidos/          .git/           setup.py        
Dockerfile      eidos.egg-info/ README.md       .travis.yml     
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ eidos -c eidos/data/meerkat_coeff_dict.npy 
usage: eidos [-h] -p PIXELS [-d DIAMETER] -f FREQ [FREQ ...]
             [-c COEFFICIENTS_FILE] [-P PREFIX] [-o8] [-N]
eidos: error: argument -p/--pixels is required
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ eidos -c eidos/data/meerkat_coeff_dict.npy -p 1024usage: eidos [-h] -p PIXELS [-d DIAMETER] -f FREQ [FREQ ...]
             [-c COEFFICIENTS_FILE] [-P PREFIX] [-o8] [-N]
eidos: error: argument -f/--freq is required
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$ eidos -c eidos/data/meerkat_coeff_dict.npy -p 1024 -f 1024
Saved as meerkat_pb_jones_cube_1channels.fits
(eidos) sperkins@simon-Y700:~/work/ska/code/eidos$
```

**Comment by JoshVStaden** [comment]
Thanks Simon. I managed to get Eidos running.

In terms of the coefficients themselves, I understand that there is a  
set of coefficients for each antenna and each frequency bin. However,  
in terms of matching coefficients to certain Zernicke terms, I noticed  
that, in Eidos, they stored the coefficients as a list with each  
element containing a number (representing the Noll index for the  
specific polynomial) and the coefficient itself along with it. This  
does make sense for the sake of having multiple polynomials describing  
a single telescope. Should I do something similar to this in my code?

Kind regards
Josh


Quoting Simon Perkins <notifications@github.com>:

**Comment by sjperkins** [comment]
@JoshVStaden How about a separate `noll_index array` (see updated docstring below)?

```python
@numba.jit(nopython=True, nogil=True, cache=True)
def zernicke_dde(coords, coeffs):
"""
Evaluate Zernicke Polynomials defined by coefficients ``coeffs`
at the specified coordinates ``coords``.

Parameters
----------
coords : :class:`numpy.ndarray`
   Coordinates at which to evaluate the zernicke polynomials.
   Has shape :code:`(3, source, time, ant, chan)`. The three components in
   the first dimension represent l, m and frequency coordinates, respectively.
coeffs : :class:`numpy.ndarray`
  Zernicke polynomial coefficients.
  Has shape :code:`(ant, chan, poly)` where ``poly`` is the number of
  polynomical coefficients.
noll_index : :class:`numpy.ndarray`
  Noll index associated with each polynomial coefficient.
  Has shape :code:`(ant, chan, poly)`.

Returns
-------
:class:`numpy.ndarray`
   complex values with shape :code:`(source, time, ant, chan)`
"""
```

**Comment by sjperkins** [comment]
@JoshVStaden Nice work! I'm trying to think of a way to avoid submitting binary data into the repo. Can you think of a way to modify your test cases to work without binary data?

Perhaps working with a subset of the coefficients?

**Comment by JoshVStaden** [comment]
@sjperkins I can copy and paste the coefficients into the file, if that's more preferable?

File: tests/test_zernicke.py [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+       -0.0428337 -0.03446249j,  0.07525696+0.03761065j,
+       -0.00754467+0.00811033j,  0.01189913+0.01875151j,
+        0.00248063+0.00179074j,  0.00160786+0.00614232j,
+       -0.01133655-0.01143651j,  0.00470805-0.01920698j,
+        0.0038768 -0.00601548j,  0.00172058-0.00385759j,
+       -0.01082336-0.00432746j, -0.0009297 +0.00796986j,
+        0.01785803+0.00319331j])
+
+@pytest.fixture
+def index_xx():
```
Comment (sjperkins) [inline tests/test_zernicke.py]: Rename to `noll_index_xx` and similarly for the other 3 correlations.

File: africanus/rime/zernicke.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import numba
+import numpy as np
+import math
+
+@numba.jit(nogil=True, nopython=True)
```
Comment (sjperkins) [inline africanus/rime/zernicke.py]: Add a `cache=True` to the jit decorator and all those below.

**Comment by sjperkins** [comment]
@JoshVStaden Thanks for doing that.

You need to merge the current master branch into your branch and resolve the conflicts before this entire branch can be merged.

**Comment by JoshVStaden** [comment]
@sjperkins I just resolved the conflicts. The merge should be done by now.

**Comment by sjperkins** [comment]
@JoshVStaden Thanks, it looks like the test cases are failing. Could you please resolve this?

See https://github.com/ska-sa/codex-africanus/pull/18#issue-183281289 for more information.

**Comment by sjperkins** [comment]
@JoshVStaden Still failing.

I think some imports got left out during the merge.

See https://travis-ci.org/ska-sa/codex-africanus/builds/392148291 for more details.

**Comment by JoshVStaden** [comment]
@sjperkins I think so. There is a function called dask_xform_src and one called np_xform_src that does not seem to be defined anywhere in the files. On top of that, though, it does not seem to be imported from anywhere in particular. Is there a particular file that it is meant to be imported from, or a package that is meant to be included?

**Comment by sjperkins** [comment]
@JoshVStaden Take a look at the "File changed" tab on the top of this PR to see how they were imported in master brach

**Comment by JoshVStaden** [comment]
@sjperkins All checks seem to be passing now.

**Comment by sjperkins** [comment]
Excellent! Last two minor points

- Add yourself to AUTHORS.rst
- Add this PR to HISTORY.rst

File: docs/api.rst [hunk 3]
```diff
@@ -11,3 +11,57 @@ API
...
+------------------------
+
+.. autofunction:: africanus.deconv.hogbom.hogbom_clean
+
+Utilities
+---------
+
+.. autofunction:: africanus.util.beams.beam_filenames
+.. autofunction:: africanus.util.beams.beam_grids
+
```
Comment (sjperkins) [inline docs/api.rst]: Wait, also remove all these additions from `api.rst`. They  now live in `rime-api.rst` where I see that you've correctly added the `zernicke` function.

**Comment by sjperkins** [comment]
Merged

**Pull request merged**